# Analyzing borrowers’ risk of defaulting

This analysis is based on data taken from a major bank loan division, in order to determine if customer’s marital status and number of children has an impact on the ability of a potential borrower to repay their loan on time.

The project explains how to do simple Data Preprocessing in Python:

1. Working with Missing Values
2. Changing Data Types
3. Looking for Duplicates
4. Categorizing Data

#### Importing libraries and data

In [1]:
import pandas as pd
data = pd.read_csv("C:\\Users\\Shir\\Downloads\\Practicum\\Library\\credit_scoring_eng.csv")

data.info()
data.describe()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     19351 non-null  float64
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      19351 non-null  float64
 11  purpose           21525 non-null  object 
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


children  days_employed     dob_years  education_id  \
count  21525.000000   19351.000000  21525.000000  21525.000000   
mean       0.538908   63046.497661     43.293380      0.817236   
std        1.381587  140827.311974     12.574584      0.548138   
min       -1.000000  -18388.949901      0.000000      0.000000   
25%        0.000000   -2747.423625     33.000000      1.000000   
50%        0.000000   -1203.369529     42.000000      1.000000   
75%        1.000000    -291.095954     53.000000      1.000000   
max       20.000000  401755.400475     75.000000      4.000000   

       family_status_id          debt   total_income  
count      21525.000000  21525.000000   19351.000000  
mean           0.972544      0.080883   26787.568355  
std            1.420324      0.272661   16475.450632  
min            0.000000      0.000000    3306.762000  
25%            0.000000      0.000000   16488.504500  
50%            0.000000      0.000000   23202.870000  
75%            1.000000      0.000000   32549.611000  
max            4.000000      1.000000  362496.645000

In [2]:
#checking for duplicates and errors
data.nunique()

children                8
days_employed       19351
dob_years              58
education              15
education_id            5
family_status           5
family_status_id        5
gender                  3
income_type             8
debt                    2
total_income        19348
purpose                38
dtype: int64

In [3]:
#exploring the gender column
data['gender'].value_counts()

F      14236
M       7288
XNA        1
Name: gender, dtype: int64

In [4]:
#exploring the education column
data['education'].unique()

array(["bachelor's degree", 'secondary education', 'Secondary Education',
       'SECONDARY EDUCATION', "BACHELOR'S DEGREE", 'some college',
       'primary education', "Bachelor's Degree", 'SOME COLLEGE',
       'Some College', 'PRIMARY EDUCATION', 'Primary Education',
       'Graduate Degree', 'GRADUATE DEGREE', 'graduate degree'],
      dtype=object)

### Conclusion

There are some missing values, errors and imperfect data that we will need to address:

1. Missing values in 2 columns(days_employed and total_income) - quantitative variables.
2. Illogical values (for example: children = 20,-1,  dob_years = 0).
3. Errors - negetive values in days_employed column.
4. Gender column should have only 2 uniqe values (Male(M) and Female(F)), table now show 3 values('XNA' not defined).
5. Duplicate values in education column(becaue of case sensitivity), number of uniqe values in education column should be equal to the number of uniqe values in education_id column.      


## Data preprocessing

### Processing missing values

In [6]:
#Check the number of rows
print('Total Rows:', len(data)) 


# look for missing values per column
print(data.isnull().sum())

Total Rows: 21525
children               0
days_employed       2174
dob_years              0
education              0
education_id           0
family_status          0
family_status_id       0
gender                 0
income_type            0
debt                   0
total_income        2174
purpose                0
dtype: int64


In [7]:
#Replacing missingvalues with 0
data['days_employed'] = data['days_employed'].fillna(value= 0)
data['total_income'] = data['total_income'].fillna(value= 0)
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     21525 non-null  float64
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      21525 non-null  float64
 11  purpose           21525 non-null  object 
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


### Conclusion

There are 21525 rows, missing values were filled by replacing the missing values with 0.
Possible explanation for the the missing values: no value means the customer is unemployed, and have no income and no days employed.

### Data type replacement

In [8]:
#Replacing negetive values in the column days_employed with their positive value(days_employed can't be negative)
data['days_employed'] =  data['days_employed'].abs()

#test that the change was made
data.describe()

children  days_employed     dob_years  education_id  \
count  21525.000000   21525.000000  21525.000000  21525.000000   
mean       0.538908   60156.419005     43.293380      0.817236   
std        1.381587  133355.929525     12.574584      0.548138   
min       -1.000000       0.000000      0.000000      0.000000   
25%        0.000000     610.652074     33.000000      1.000000   
50%        0.000000    1808.053434     42.000000      1.000000   
75%        1.000000    4779.587738     53.000000      1.000000   
max       20.000000  401755.400475     75.000000      4.000000   

       family_status_id          debt   total_income  
count      21525.000000  21525.000000   21525.000000  
mean           0.972544      0.080883   24082.055063  
std            1.420324      0.272661   17583.554088  
min            0.000000      0.000000       0.000000  
25%            0.000000      0.000000   14178.053000  
50%            0.000000      0.000000   21682.354000  
75%            1.000000      0.000000   31286.979000  
max            4.000000      1.000000  362496.645000

In [9]:
# There are approximately twice more female then male, so 'XNA' value in the gender column will be replaced with the most common value 'F'
data.loc[data['gender'] == 'XNA', ['gender'] ] = 'F'

#test that changes were made and explore the 'gender column:
data['gender'].value_counts()

F    14237
M     7288
Name: gender, dtype: int64

In [10]:
#Exploring the children column values to identify strange values:
print(data.groupby(['children']).count())

          days_employed  dob_years  education  education_id  family_status  \
children                                                                     
-1                   47         47         47            47             47   
 0                14149      14149      14149         14149          14149   
 1                 4818       4818       4818          4818           4818   
 2                 2055       2055       2055          2055           2055   
 3                  330        330        330           330            330   
 4                   41         41         41            41             41   
 5                    9          9          9             9              9   
 20                  76         76         76            76             76   

          family_status_id  gender  income_type   debt  total_income  purpose  
children                                                                       
-1                      47      47           47     47     

In [11]:
#Replacing min and max values(20 and -1, values are illogical) by replacing 20 with 2 and -1 with 1
data.loc[data['children'] == 20, ['children'] ] = 2
data.loc[data['children'] == -1, ['children'] ] = 1

#test that the change was made
print(data.groupby(['children']).count())

          days_employed  dob_years  education  education_id  family_status  \
children                                                                     
0                 14149      14149      14149         14149          14149   
1                  4865       4865       4865          4865           4865   
2                  2131       2131       2131          2131           2131   
3                   330        330        330           330            330   
4                    41         41         41            41             41   
5                     9          9          9             9              9   

          family_status_id  gender  income_type   debt  total_income  purpose  
children                                                                       
0                    14149   14149        14149  14149         14149    14149  
1                     4865    4865         4865   4865          4865     4865  
2                     2131    2131         2131   2131 

In [12]:
# calculating the avergare age and replacing '0' values in dob_years column with the average age, as age can't be 0
data.loc[data['dob_years'] == 0, ['dob_years']] = data['dob_years'].mean()


data.describe()

children  days_employed     dob_years  education_id  \
count  21525.000000   21525.000000  21525.000000  21525.000000   
mean       0.479721   60156.419005     43.496522      0.817236   
std        0.755528  133355.929525     12.218174      0.548138   
min        0.000000       0.000000     19.000000      0.000000   
25%        0.000000     610.652074     34.000000      1.000000   
50%        0.000000    1808.053434     43.000000      1.000000   
75%        1.000000    4779.587738     53.000000      1.000000   
max        5.000000  401755.400475     75.000000      4.000000   

       family_status_id          debt   total_income  
count      21525.000000  21525.000000   21525.000000  
mean           0.972544      0.080883   24082.055063  
std            1.420324      0.272661   17583.554088  
min            0.000000      0.000000       0.000000  
25%            0.000000      0.000000   14178.053000  
50%            0.000000      0.000000   21682.354000  
75%            1.000000      0.000000   31286.979000  
max            4.000000      1.000000  362496.645000

### Conclusion

1. days_employed column - negetive values in columns days_employed were replaced with their positive value.
2. gender column - 'XNA' value was replaced with the most common value ('F').
3. children column - min and max values(20 and -1) were replaced: 20 with 2 and -1 with 1.
4. dob_years column - values equal to 0 were replaced with the average age.

### Processing duplicates

In [13]:
#exploring the values in education column
data.pivot_table(index='education', values= 'education_id', aggfunc='count')

education_id
education                        
BACHELOR'S DEGREE             274
Bachelor's Degree             268
GRADUATE DEGREE                 1
Graduate Degree                 1
PRIMARY EDUCATION              17
Primary Education              15
SECONDARY EDUCATION           772
SOME COLLEGE                   29
Secondary Education           711
Some College                   47
bachelor's degree            4718
graduate degree                 4
primary education             250
secondary education         13750
some college                  668

In [14]:
#count the number of uniqe values in the education columns
print('Unique values:', data['education'].nunique())

#count the number of values in education column
print('Total values:', len(data['education']))


Unique values: 15
Total values: 21525


In [15]:
#remove case sensitivity duplicates using str.lower()
data['education'] = data['education'].str.lower()

In [16]:
#exploring the values in education column after the change
data.pivot_table(index='education', values= 'education_id', aggfunc='count')

education_id
education                        
bachelor's degree            5260
graduate degree                 6
primary education             282
secondary education         15233
some college                  744

In [17]:
#count the number of uniqe values in the education columns after the change
print('Unique values:', data['education'].nunique())

#count the number of values in education column after the change
print('Total values:', len(data['education']))

Unique values: 5
Total values: 21525


### Conclusion

Duplicates in education column were removed, now the number of uniqe values in education column match the numbner of uniqe values in the educaton_id column

### Categorizing Data

In [18]:
#create 2 new df with the groups of people who defaulted on a loan, and people who don't - it will help when we would want to compare between groups

data_debt = data[data['debt'] == 1]
data_no_debt = data[data['debt'] == 0]


In [19]:
#look at the reasons for taking out a loan, and try to see any trends
data.purpose.value_counts()

wedding ceremony                            797
having a wedding                            777
to have a wedding                           774
real estate transactions                    676
buy commercial real estate                  664
housing transactions                        653
buying property for renting out             653
transactions with commercial real estate    651
purchase of the house                       647
housing                                     647
purchase of the house for my family         641
construction of own property                635
property                                    634
transactions with my real estate            630
building a real estate                      626
buy real estate                             624
building a property                         620
purchase of my own house                    620
housing renovation                          612
buy residential real estate                 607
buying my own car                       

In [20]:
#There are few trends: cars, wedding and education
#count how many times people used each purpose in debt data(debt = 1):
print('Trend: real estate' , sum("real estate" in s for s in data_debt['purpose']))
print('Trend: house' , sum("hous" in s for s in data_debt['purpose']))
print('Trend: property' , sum("property" in s for s in data_debt['purpose']))
print('Trend: car' , sum('car' in s for s in data_debt['purpose']))
print('Trend: education' , sum('educat' in s for s in data_debt['purpose']))
print('Trend: wedding' , sum('wedding' in s for s in data_debt['purpose']))
print('Trend: university' , sum('university' in s for s in data_debt['purpose']))

Trend: real estate 336
Trend: house 256
Trend: property 190
Trend: car 403
Trend: education 327
Trend: wedding 186
Trend: university 83


In [21]:
#add the similar purposes to create 4 main trends, and their total:
#use debt data to calculate only people who defaulted on a loan
wedding_debt = sum('wedding' in s for s in data_debt['purpose'])
print('wedding_debt', wedding_debt)
education_debt = ((sum('university' in s for s in data_debt['purpose'])) + sum('educat' in s for s in data_debt['purpose']))
print('education_debt', education_debt)
car_debt = (sum('car' in s for s in data_debt['purpose']))
print('car_debt', car_debt)
real_estate_debt = (sum("real estate" in s for s in data_debt['purpose']) + sum("hous" in s for s in data_debt['purpose']) + sum("property" in s for s in data_debt['purpose']))
print('real_estate_debt', real_estate_debt)
total_purpose_debt = wedding_debt+ education_debt + car_debt + real_estate_debt
print('total_purpose_debt', total_purpose_debt)

wedding_debt 186
education_debt 410
car_debt 403
real_estate_debt 782
total_purpose_debt 1781


In [22]:
#There are few trends: cars, wedding and education
#count how many times people used each purpose in no debt data(debt = 0):
print('Trend: real estate' , sum("real estate" in s for s in data_no_debt['purpose']))
print('Trend: house' , sum("hous" in s for s in data_no_debt['purpose']))
print('Trend: property' , sum("property" in s for s in data_no_debt['purpose']))
print('Trend: car' , sum('car' in s for s in data_no_debt['purpose']))
print('Trend: education' , sum('educat' in s for s in data_no_debt['purpose']))
print('Trend: wedding' , sum('wedding' in s for s in data_no_debt['purpose']))
print('Trend: university' , sum('university' in s for s in data_no_debt['purpose']))

Trend: real estate 4142
Trend: house 3564
Trend: property 2352
Trend: car 3912
Trend: education 3199
Trend: wedding 2162
Trend: university 866


In [23]:
#add the similar purposes to create 4 main trends, and their total:
#use no debt data to calculate only people who did not defaulted on a loan
wedding = sum('wedding' in s for s in data_no_debt['purpose'])
print('Wedding', wedding)
education = ((sum('university' in s for s in data_no_debt['purpose'])) + sum('educat' in s for s in data_no_debt['purpose']))
print('Education', education)
car = (sum('car' in s for s in data_no_debt['purpose']))
print('Car', car)
real_estate = (sum("real estate" in s for s in data_no_debt['purpose']) + sum("hous" in s for s in data_no_debt['purpose']) + sum("property" in s for s in data_no_debt['purpose']))
print('Real_estate', real_estate)
total_purpose = wedding+ education + car + real_estate
print('Total', total_purpose)

Wedding 2162
Education 4065
Car 3912
Real_estate 10058
Total 20197


In [24]:
#look at the median

print(data['total_income'].median())

21682.354


In [25]:
# create groups by income levels

def income_level(income):
    if income <= 21682*0.5:
        return 'low'
    if income >= 21682*1.5:
        return 'high'
    return 'medium'

#add a column with the income level
data['income_level'] = data['total_income'].apply(income_level)

In [26]:
#look at the distribuation:
data.income_level.value_counts()

medium    13176
high       4852
low        3497
Name: income_level, dtype: int64

In [27]:
data.pivot_table(index= 'income_level', columns= 'debt', values = 'total_income', aggfunc='count')

debt              0     1
income_level             
high           4511   341
low            3234   263
medium        12039  1137

In [28]:
# Categorize by people who have children, and people who don't

def have_children(children):
    if children == 0:
        return 'No'
    if children > 0:
        return 'Yes'
    return Nan

#add a column with the income level
data['have_children'] = data['children'].apply(have_children)

In [29]:
#look at the distribuation:
data.have_children.value_counts()

No     14149
Yes     7376
Name: have_children, dtype: int64

### Conclusion

More Female defaulted on a loan than male.
People with secondary education are more likely to defaulte on a loan.
Married people are more likely to defaulte on a loan.

Total income was grouped to 3 income levels, where low level was set to be 0.5 of the median, and high was set to 1.5 of the median, rest was set to medium.

Children were grouped by creating new column to detemine if the person have children or don't.


### Is there a relation between having kids and repaying a loan on time?

In [30]:
#Exploring people with defaulted loans and the number of children they have
data_debt['children'].value_counts()

0    1063
1     445
2     202
3      27
4       4
Name: children, dtype: int64

In [31]:
#calculate the percentage of people with 0 children who defaulted on a loan out of the total number of people who defaulted
data_debt.loc[data_debt.children == 0, 'children'].count() / data_debt['children'].count()

0.6105686387133831

In [32]:
#Explore people with no defaultes and the number of children they have
data_no_debt['children'].value_counts()

0    13086
1     4420
2     1929
3      303
4       37
5        9
Name: children, dtype: int64

In [33]:
#calculate the percentage of people with 0 children out of no debt data base
data_no_debt.loc[data_no_debt.children == 0, 'children'].count() / data_no_debt['children'].count()


0.6614435907804286

### Conclusion

Although it seems that 61% of the people who defaulted on a loan have 0 children, it is similar to the total percent of people in the database, so we can't assume there is a relation. 

### Is there a relation between marital status and repaying a loan on time?

In [34]:
data_no_debt['family_status'].value_counts()

married              11449
civil partnership     3789
unmarried             2539
divorced              1110
widow / widower        897
Name: family_status, dtype: int64

In [35]:
#calculate the percentage of married people out of all people who did not defaulte on a loan
data_no_debt.loc[data_no_debt.family_status == 'married', 'family_status'].count() / data_no_debt['family_status'].count()

0.5786999595632835

In [36]:
data_debt['family_status'].value_counts()

married              931
civil partnership    388
unmarried            274
divorced              85
widow / widower       63
Name: family_status, dtype: int64

In [37]:
#calculate the percentage of married people out of all people defaulted on a loan
data_debt.loc[data_debt.family_status == 'married', 'family_status'].count() / data_debt['family_status'].count()

0.5347501435956347

### Conclusion

Again, it seems like married people are more likely to defaulte on a loan, but when exploring the same status with people who didn't default on a loan, it seems like the percent is almost the same.


### Is there a relation between income level and repaying a loan on time?

In [38]:
#calculate the average income of people who are not repaying a loan on time
data_debt['total_income'].mean()

23547.984776565205

In [39]:
#calculate the average income of people repaying loans on time
data_no_debt['total_income'].mean()

24129.05346416292

### Conclusion

Both groups, people who defaulted on loans and people who didn't, have similar average income, and it's seems like there is no relation.

### How do different loan purposes affect on-time repayment of the loan?

In [40]:
#calculating the ratio between people who used aeach purpose:
print('Debt = 1 (defaulted on a loan)')
print( '-------------')
print('real_estate', real_estate_debt/ total_purpose_debt)
print('education', education_debt/ total_purpose_debt)
print('car',car_debt/ total_purpose_debt)
print('wedding', wedding_debt/ total_purpose_debt)
print('')
print('Debt = 0 (did not default on a loan)')
print( '-------------')
print('real', real_estate/ total_purpose)
print('education', education/ total_purpose)
print('car', car/ total_purpose)
print('wedding', wedding/ total_purpose)

Debt = 1 (defaulted on a loan)
-------------
real_estate 0.43907916900617633
education 0.23020774845592365
car 0.22627737226277372
wedding 0.10443571027512634

Debt = 0 (did not default on a loan)
-------------
real 0.4979947516957964
education 0.2012675149774719
car 0.19369213249492498
wedding 0.1070456008318067


### Conclusion

We see that people who are taking a loan for education or car reasons, are more likely to default on a loan, as their percent in the debt category is slightly higher then theire percent in the general, non debt category.

## General conclusion

We learned that there are some characteristics to people who are more likely to default a loan, but the information we got from the bank, can't really point to a clear relation between specific factors.
The data needs to be explored more, to undersatnd what factors do effect the chance to default a loan.

Data was imported, examined and data preprocessing was made by replacing missing values for unemployed people and illogical values regarding number of children. Negative values coused by errors were replaced by positive values, and duplicates that were created because of case sensitivity was removed. Loan purposes were categorized and sorted by main trends, and new data frames were created in order compare both groups(debt vs. no debt).

Main findings:
1. The profile of a person who will likely default on a loan is someone married with 0 children, with an avergae income of 23.5K, asking for a loan for purposes realted to real estate.

2. After comparing the group of people that defaulted on a loan to the group of people who did not, we see that both groups have similar atributes, with few small differnces:
people who are taking a loan for education or car related reasons, are more likely to default on a loan.
